Course Work, check it, pls

Spolpinka opened the pull request:
Курсовая, уровень "очень сложно", прошу проверить.

Spolpinka committed:
Course Work, check it, pls

dmitry-bizin commented on src/Employee.java:
```diff
@@ -0,0 +1,53 @@
+    private double salary;
+    private int id;
+
+    Employee(int id, String fullname, int department, double salary) {
```
id не нужно передавать в конструктор, а нужно получать в конструкторе из статической переменной-счётчика (см. базовый уровень)

dmitry-bizin commented on src/Employee.java:
```diff
@@ -0,0 +1,53 @@
+        this.department = department;
+    }
+
+    public void setFullname(String fullname) {
```
этот сеттер не нужен по заданию

dmitry-bizin commented on src/Employee.java:
```diff
@@ -0,0 +1,53 @@
+        this.fullname = fullname;
+    }
+
+    public void setId(int id) {
```
этот сеттер не нужен по заданию

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+import java.util.Arrays;
+
+public class EmployeeBook {
+    static private Employee[] employees = new Employee[10];
```
это поле не должно быть statiic, и лучше сделать его final

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+    static private Employee[] employees = new Employee[10];
+
+    //getters
+    Employee[] getEmployees() {
```
геттер не нужен

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        for (Employee emp :
+                EmployeeBook.employees
+        ) {
+            sum += emp.getSalary();
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        double max = 0;
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getSalary() >= max) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        double min = EmployeeBook.employees[0].getSalary();
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getSalary() <= min) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+
+    Employee getMinSalary() {
+        Employee minSalary = null;
+        double min = EmployeeBook.employees[0].getSalary();
```
а если не будет сотрудника на 0 месте? вот создам объект EmployeeBook и сразу вызову этот метод и будет ошибка

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+    void getHighSalary(double percent) {
+        for (Employee emp :
+                EmployeeBook.employees) {
+            emp.setSalary(emp.getSalary() + emp.getSalary() * percent / 100);
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        Employee minSalary = null;
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getDepartment() == x && min > emp.getSalary()) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    Employee getMinSalaryInDep(int x) {
+        double min = EmployeeBook.employees[0].getSalary();
```
а если не будет сотрудника на 0 месте? вот создам объект EmployeeBook и сразу вызову этот метод и будет ошибка

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        Employee maxSalary = null;
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getDepartment() == x && max < emp.getSalary()) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        double sum = 0;
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getDepartment() == x) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        int numberEmpInDepart = 0;
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getDepartment() == x) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+    void getHighSalaryOfDepart(int x, double percent) {
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getDepartment() == x) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        }
+    }
+
+    void getHighSalaryOfDepart(int x, double percent) {
```
почему метод называется get, но ничего не возвращает?

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+
+    //setters
+    void setEmployee(int id, String fullname, int department, double salary) {
+        employees[id] = new Employee(id, fullname, department, salary);
```
id - это не индекс, этот метод не нужен. Не нужно работать с индексами массива

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+    void printAllFullnames() {
+        for (Employee emp :
+                EmployeeBook.employees) {
+            System.out.println(emp.getFullname());
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+    void printEmployeesOfDepart(int x) {
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getDepartment() == x) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        System.out.println("Следующие сотрудники получают меньше, чем " + salary);
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getSalary() < salary) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        System.out.println("Следующие сотрудники получают больше или равно, чем " + salary);
+        for (Employee emp :
+                EmployeeBook.employees) {
+            if (emp.getSalary() >= salary) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+            numbersOfDepartment[i] = employees[i].getDepartment();
+        }
+        Arrays.sort(numbersOfDepartment);
+        ArrayList depart = new ArrayList<Integer>();
```
`List<Integer> depart = new ArrayList<>();`

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        for (int i = 0; i < depart.size(); i++) {
+            System.out.println("Сотрудники отдела № " + depart.get(i));
+            for (int j = 0; j < employees.length; j++) {
+                if (i == employees[j].getDepartment()){
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+    //others methods
+    void searchDelete(String fullName, int id) {
+        for (int i = 0; i < employees.length; i++) {
+            if (employees[i].getFullname().equals(fullName) || employees[i].getId() == id) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+        }
+    }
+
+    void addEmployee(int id, String name) {
```
id не нужно передавать, он присваивается в конструкторе Employee

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,235 @@
+    void changeEmployeeSalaryDepartment(String fullName, double newSalary, int newDepartment) {
+        for (Employee emp :
+                employees) {
+            if (emp.getFullname().equals(fullName)) {
```
а если не будет сотрудника в массиве? будет исключение. Нужно добавить проверку на налл

dmitry-bizin commented on src/Main.java:
```diff
@@ -1,5 +1,133 @@
+        return random.nextInt(100_000) + 100_000;
+    }
+
+    static int getRandomDepartment() {
```
private

dmitry-bizin commented on src/Main.java:
```diff
@@ -1,5 +1,133 @@
+        employeeBook.printEmployeesByDeparment();
+    }
+
+    static int getRandomSalary() {
```
private

dmitry-bizin commented on src/Main.java:
```diff
@@ -1,5 +1,133 @@
 public class Main {
+    static int counter = 0;
+
+    public static void setCounter(int counter) {
```
private

dmitry-bizin commented on src/Main.java:
```diff
@@ -1,5 +1,133 @@
+import java.util.Random;
+
 public class Main {
+    static int counter = 0;
```
private

dmitry-bizin replied:
этот счётчик должен быть в Employee

Spolpinka committed:
counter перенес в Employee
во всех геттерах, где было надо, добавлены проверки на null, как у массива, так и у отдельных сотрудников
ну и остальное

dmitry-bizin commented on src/Employee.java:
```diff
@@ -0,0 +1,52 @@
+        this.department = department;
+    }
+
+    int counterGrow(int counter){
```
а этот метод зачем? не нужно давать управлять id вне класса. Счётчик - внутренняя штука класса и не более

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+
+    double getSumOfSalaries() {
+
+        if (employees != null) {
```
эта проверка смысла не имеет, т.к. массив никогда не будет налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    Employee getMaxSalaryName() {
+        if (employees != null) {
```
эта проверка смысла не имеет, т.к. массив никогда не будет налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    Employee getMinSalary() {
+        if (employees != null) {
```
эта проверка смысла не имеет, т.к. массив никогда не будет налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+    Employee getMinSalary() {
+        if (employees != null) {
+            Employee minSalary = null;
+            double min = employees[0].getSalary();
```
employees[0] может быть налл. Вот создай объект этого класса и вызови сразу этот метод - упадёт программа))

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+
+    void raiseSalary(double percent) {
+
+        if (employees != null) {
```
эта проверка смысла не имеет, т.к. массив никогда не будет налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    Employee getMinSalaryInDep(int x) {
+        if (employees != null) {
```
эта проверка смысла не имеет, т.к. массив никогда не будет налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+
+    Employee getMinSalaryInDep(int x) {
+        if (employees != null) {
+            double min = employees[0].getSalary();
```
здесь тоже может не быть сотрудника на 0 месте

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    Employee getMaxSalaryInDep(int x) {
+        if (employees != null) {
```
эта проверка смысла не имеет, т.к. массив никогда не будет налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    double getSumSalaryOfDepart(int x) {
+        if (employees != null) {
```
эта проверка смысла не имеет, т.к. массив никогда не будет налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    double getAverageSalaryOfDepart(int x) {
+        if (employees != null) {
```
эта проверка смысла не имеет, т.к. массив никогда не будет налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+    }
+
+    void raiseSalaryOfDepart(int x, double percent) {
+        if (employees != null) {
```
эта проверка смысла не имеет, т.к. массив никогда не будет налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+    void printAllFullnames() {
+        for (Employee emp :
+                employees) {
+            System.out.println(emp.getFullname());
```
нет проверки на налл

dmitry-bizin commented on src/EmployeeBook.java:
```diff
@@ -0,0 +1,295 @@
+        for (int i = 0; i < depart.size(); i++) {
+            System.out.println("Сотрудники отдела № " + depart.get(i));
+            for (int j = 0; j < employees.length; j++) {
+                if (i == employees[j].getDepartment()) {
```
здесь тоже нужно добавить проверку на налл
